### PR TITLE
Create custom attributes

### PR DESCRIPTION
**Pull request by tlein**
Part 1 of allowing the creation and usage of custom attributes with the editor.

This pull requests allows the creation of custom attribute templates from within the editor and saves them to a metadata file.

### CODE REVIEW

File: src/duckling/project/project.ts [hunk 1]
```diff
@@ -2,11 +2,16 @@ import {Action} from '../state';
 import {immutableAssign} from '../util';
 import {Vector} from '../math/vector';
 
+export interface CustomAttribute {
+    name: string,
```
Comment (tlein) [inline src/duckling/project/project.ts:8]: The type of this should be AttributeKey

Comment (tlein) [reply src/duckling/project/project.ts:8]: Name should be "key" not "name"

File: src/duckling/project/custom-attributes-depot.component.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                        class="dk-inline"
+                        [value]="curAddingName"
+                        [dividerColor]="isValidAttributeName(curAddingName) ? 'primary' : 'warn'"
+                        (inputChanged)="onAttributeNameInput($event)">
+                    </dk-input>
+                </div>
+                <md-list>
+                    <md-list-item
+                        *ngFor="let attribute of customAttributes"
+                        (click)="selectAttribute(attribute)">
```
Comment (tlein) [inline src/duckling/project/custom-attributes-depot.component.ts]: selectAttribute no longer exists, this needs to be removed

File: src/duckling/project/project.ts (hunk 1 above)
Comment (tlein) [reply src/duckling/project/project.ts:8]: fixed both

File: src/duckling/project/custom-attributes-depot.component.ts (hunk 2 above)
Comment (tlein) [reply src/duckling/project/custom-attributes-depot.component.ts]: fixed

File: src/duckling/project/json-schema.component.ts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+export interface JsonSchema {
+    keys: string[],
+    contents: JsonSchemaValue[]
+}
+
+export type JsonSchemaValue = "number" | "string" | "boolean" | JsonSchema | JsonSchema[];
+
+@Component({
+    selector: "dk-json-schema",
```
Comment (tlein) [inline src/duckling/project/json-schema.component.ts]: Should rename to json-edit-schema.component

File: src/duckling/project/custom-attributes-depot.component.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import {CustomAttribute} from './custom-attribute';
+import {JsonSchema, schemaToJson} from './json-schema.component';
+
+
+/**
+ * Component where the user can add, delete, and edit custom attributes
+ */
+@Component({
+    selector: "dk-custom-attributes-depot",
+    styleUrls: ["./duckling/layout.css", "./duckling/project/custom-attributes-depot.component.css"],
```
Comment (jeffswenson) [inline src/duckling/project/custom-attributes-depot.component.ts]: The UX of this screen should be changed to resemble the splash screen.

File: src/duckling/project/custom-attributes-depot.component.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                        Not currently adding a new attribute
+                    </div>
+                </div>
+                <div class="adding-section-footer">
+                    <button 
+                        md-raised-button
+                        color="primary"
+                        [disabled]="!isValidAttributeName(curAddingName)"
+                        (click)="onSaveCustomAttribute()">
+                        Save
```
Comment (jeffswenson) [inline src/duckling/project/custom-attributes-depot.component.ts]: This doesn't actually save. You should change the name to Accept.

File: src/duckling/project/custom-attributes-depot.component.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                    <md-list-item *ngFor="let attribute of customAttributes">
+                        {{attribute.key}}
+                    </md-list-item>
+                </md-list>
+            </dk-section>
+            <dk-section 
+                class="adding-section"
+                [headerText]="'Currently adding: ' + addingSectionHeader">
+                <div class="adding-section-body">
+                    <div *ngIf="curAddingSchema !== undefined">
```
Comment (jeffswenson) [inline src/duckling/project/custom-attributes-depot.component.ts]: I don't think you should use curAddingSchema !== undefined as a state test. You should manage the state more explicitly.

File: src/duckling/project/project.service.ts [hunk 7]
```diff
@@ -48,9 +49,22 @@ export class ProjectService {
...
     }
 
+    private _loadProjectMetaData() {
+        this._loadCustomAttributes();
+    }
+
+    private async _loadCustomAttributes() {
+        let attributes = await glob(`${this._customAttributesRoot}/*.json`);
+        for (let customAttribute of attributes) {
+            let json = await this._jsonLoader.getJsonFromPath(customAttribute);
```
Comment (jeffswenson) [inline src/duckling/project/project.service.ts]: Changing this to Promise.map allows you to load in parallel.

File: src/duckling/project/json-schema.component.ts [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+        './duckling/project/json-schema.component.css'
+    ],
+    template: `
+        <div class="json-content">
+            {{ '{' }}
+            <div
+                *ngFor="let index of getKeyIndices()"
+                class="json-key">
+                <div [ngSwitch]="jsonType(jsonSchema.contents[index])">
+                    <div *ngSwitchCase="JsonValueType.Number">
```
Comment (jeffswenson) [inline src/duckling/project/json-schema.component.ts]: I think this could be dryer.

**Review by jeffswenson — CHANGES_REQUESTED** [review]

File: src/duckling/project/custom-attributes-depot.component.ts (hunk 4 above)
Comment (tlein) [reply src/duckling/project/custom-attributes-depot.component.ts]: fixed

File: src/duckling/project/custom-attributes-depot.component.ts (hunk 6 above)
Comment (tlein) [reply src/duckling/project/custom-attributes-depot.component.ts]: fixed

File: src/duckling/project/custom-attributes-depot.component.ts (hunk 5 above)
Comment (tlein) [reply src/duckling/project/custom-attributes-depot.component.ts]: fixed

**Comment by tlein** [comment]
Loading in custom attributes during map load needs to clear the undo/redo stack

EDIT: Fixed

File: src/duckling/project/project.service.ts (hunk 7 above)
Comment (tlein) [reply src/duckling/project/project.service.ts]: fixed

Comment (tlein) [reply src/duckling/project/project.service.ts]: fixed

**Comment by tlein** [comment]
custom-attributes-depot should drop the depot from the name

EDIT: Fixed

File: src/duckling/project/json-schema.component.ts (hunk 8 above)
Comment (tlein) [reply src/duckling/project/json-schema.component.ts]: fixed

File: src/duckling/project/project.service.ts [hunk 9]
```diff
@@ -46,9 +47,25 @@ export class ProjectService {
...
+
+    private async _loadCustomAttributes() {
+        let attributes = await glob(`${this._customAttributesRoot}/*.json`);
+        let attributePromises : Promise<void>[] = [];
+        for (let customAttribute of attributes) {
+            attributePromises.push(this._jsonLoader.getJsonFromPath(customAttribute).then(json => {
+                this.addCustomAttribute(this._customAttributeFileToName(customAttribute), JSON.parse(json))
+            }));
+        }
+        Promise.all(attributePromises);
```
Comment (jeffswenson) [inline src/duckling/project/project.service.ts]: You need a return here.

File: src/duckling/project/custom-attributes.component.ts [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+@Component({
+    selector: "dk-custom-attributes",
+    styleUrls: ["./duckling/layout.css", "./duckling/project/custom-attributes.component.css"],
+    template: `
+        <div class="container">
+            <md-card class="left-section">
+                <md-card-content>
+                    <md-list>
+                        <md-list-item
+                            md-list-item
```
Comment (jeffswenson) [inline src/duckling/project/custom-attributes.component.ts]: You don't need the md-list-item

File: src/duckling/project/custom-attributes.component.scss [hunk 11]
```diff
@@ -0,0 +1,67 @@
+@import '../forms';
+@import '../fonts';
+@import '../colors';
+
+$background-color: #FFFFFF;
+$attribute-list-background-color: $ultra-light-primary-color;
+$attribute-list-border-color: $light-primary-color;
+
+.container {
```
Comment (jeffswenson) [inline src/duckling/project/custom-attributes.component.scss:9]: https://www.google.com/search?q=material+ui+dialogs&espv=2&biw=1893&bih=933&source=lnms&tbm=isch&sa=X&ved=0ahUKEwjR94PLwtTRAhXkxFQKHYt-DbkQ_AUIBigB&dpr=1.25#imgrc=hGhlFtGv0HbhpM%3A

File: src/duckling/project/json-schema.component.ts [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+    SimpleChange
+} from '@angular/core';
+
+import {CustomAttribute} from './custom-attribute';
+import {JsonValueType} from '../controls/json.component';
+import {immutableAssign, immutableArrayAssign, immutableArrayDelete} from '../util/model';
+import {toSelectOptions, SelectOption} from '../controls/array-select.component';
+
+export interface JsonSchema {
+    keys: string[],
```
Comment (JeffSwenson) [inline src/duckling/project/json-schema.component.ts]: I would changes this to one array and add key to the JsonSchemaValue.

**Review by jeffswenson — CHANGES_REQUESTED** [review]

File: src/duckling/project/project.service.ts (hunk 9 above)
Comment (tlein) [reply src/duckling/project/project.service.ts]: fixed

File: src/duckling/project/custom-attributes.component.ts (hunk 10 above)
Comment (tlein) [reply src/duckling/project/custom-attributes.component.ts]: fixed

File: src/duckling/project/custom-attributes.component.scss (hunk 11 above)
Comment (tlein) [reply src/duckling/project/custom-attributes.component.scss:9]: Wrote up in #234

**Review by jeffswenson — APPROVED** [review]

**Pull request merged**